src/api/app/models/workflow/step/set_flags.rb

```diff
@@ -0,0 +1,66 @@
+class Workflow::Step::SetFlags < ::Workflow::Step
+  REQUIRED_KEYS = [:flags]
+  REQUIRED_FLAG_KEYS = [:type, :status, :project].freeze
+  OPTIONAL_FLAG_KEYS = [:package, :repository, :architecture].freeze
+
+  validate :validate_project_and_package
+  validate :validate_flags
+
+  def call(options = {})
+    return unless valid?
+
+    set_flags
+  end
+
+  def flags
+    step_instructions[:flags]
+  end
+
+  private
+
+  def set_flags
+    flags.each do |flag|
+      project = Project.find_by(name: flag[:project])
+      package = if project.present? && flag[:package].present?
+                  project.packages.find_by(name: flag[:package])
+                else
+                  nil
+                end
+
+      main_object = package.present? ? package : project
+      Pundit.authorize(@main_object, :update?)
```

Comment:
I think here it shouldn't be `@main_object` but `main_object`, right?

Reply:
Oh right.